packages/hierarchies/src/hierarchies/internal/operators/Grouping.ts

```diff
@@ -88,34 +100,33 @@ async function groupInstanceNodes(
   nodes: ProcessedInstanceHierarchyNode[],
   extraSiblings: number,
   groupingHandlers: GroupingHandler[],
-  onGroupingNodeCreated: undefined | ((groupingNode: ProcessedGroupingHierarchyNode) => void),
+  parentNode: ParentHierarchyNode | undefined,
+  onGroupingNodeCreated?: (groupingNode: ProcessedGroupingHierarchyNode) => void,
 ): Promise<Array<ProcessedGroupingHierarchyNode | ProcessedInstanceHierarchyNode>> {
-  let curr: GroupingHandlerResult<ProcessedGroupingHierarchyNode> | undefined;
-  for (let i = 0; i < groupingHandlers.length; ++i) {
-    const currentHandler = groupingHandlers[i];
-    const nextHandlers = groupingHandlers.slice(i + 1);
-    const groupings = assignAutoExpand(applyGroupHidingParams(await currentHandler(curr?.ungrouped ?? nodes), extraSiblings));
+  let curr: GroupingHandlerResult | undefined;
+  for (const currentHandler of groupingHandlers) {
+    const groupings = assignAutoExpand(applyGroupHidingParams(await currentHandler(curr?.ungrouped ?? nodes, curr?.grouped ?? []), extraSiblings));
     curr = {
       groupingType: groupings.groupingType,
-      grouped: mergeSortedArrays(
-        curr?.grouped ?? [],
-        await Promise.all(
-          groupings.grouped.map(
-            async (grouping): Promise<ProcessedGroupingHierarchyNode> => ({
-              ...grouping,
-              children: await groupInstanceNodes(grouping.children, 0, nextHandlers, onGroupingNodeCreated),
-            }),
-          ),
-        ),
-        compareNodesByLabel,
-      ),
+      grouped: [...(curr?.grouped ?? []), ...groupings.grouped],
       ungrouped: groupings.ungrouped,
     };
   }
   if (curr) {
```

Comment:
We can reduce the amount of nesting here. How about this:
```ts
  if (!curr) {
    return nodes;
  }

  for (const groupingNode of curr.grouped) {
    onGroupingNodeCreated?.(groupingNode);
    if (!parentNode) {
      continue;
    }

    if (HierarchyNode.isGroupingNode(parentNode)) {
      groupingNode.nonGroupingAncestor = parentNode.nonGroupingAncestor;
      continue;
    }

    // not sure why type checker doesn't pick this up
    assert(HierarchyNode.isCustom(parentNode) || HierarchyNode.isInstancesNode(parentNode));
    groupingNode.nonGroupingAncestor = parentNode;
  }

  return [...curr.grouped, ...curr.ungrouped];
```

Reply:
agree, fixed